### ISSUE TITLE
[python] Build ARM/M1 wheels for Mac

### ISSUE THREAD

**Issue by bkmartinjr** [single-cell-data/TileDB-SOMA#794]
Currently we only build x86 wheels - we will have lots of users on Mac M1/M2 ARM CPUs

N.B. Needs to be sequenced after #494 lands

**Comment by mlin** [comment]
Yes, I believe cibuildwheel can build ARM wheels using some QEMU scheme inside GitHub Actions workers. Will look into it

**Comment by johnkerl** [comment]
Also note that TileDB core lacks M1 wheels, which is another area we can explore with the core team:
https://pypi.org/project/tiledb/0.20.0/#files

**Comment by johnkerl** [comment]
Failed attempt at #810

**Comment by johnkerl** [comment]
Also note here is the `TileDB-Py` PR we can maybe learn from:
https://github.com/TileDB-Inc/TileDB-Py/pull/1614/files

**Comment by mlin** [comment]
@atolopko-czi braindump: our [python packaging workflow](https://github.com/single-cell-data/TileDB-SOMA/blob/6e3e47d16584dfb664f151c35dcba38b45fd7385/.github/workflows/python-packaging.yml#L38-L75) uses [cibuildwheel](https://cibuildwheel.readthedocs.io/en/stable/) to build the wheels for the {platforms} x {python versions} matrix, including x86-64 macOS using the macOS GitHub Actions workers.

cibuildwheel has a [scheme to build universal macOS binaries](https://cibuildwheel.readthedocs.io/en/stable/faq/#apple-silicon) which in theory we just need to enable. This is based on cross-compiling from the x86-64 workers, though, so one needs an M1 machine with which to test them.

Above John linked to a TileDB-Py diff enabling this cross-compilation (also using cibuildwheel).

**Comment by atolopko-czi** [comment]
Got past the original failure by adding these env vars to the `pypa/cibuildwheel` step's env:
```
        CIBW_ARCHS_MACOS: ${{ matrix.cibw_archs_macos }}
        CIBW_TEST_SKIP: "*_arm64"
```
Subsequent operations `pypa/cibuildwheel` fails (`delocate-wheel`):
```
 delocate.delocating.DelocationError: Some missing architectures in wheel
  Required arch arm64 missing from tiledbsoma/libtiledbsoma.dylib
                                                              ✕ 9.07s
Error: Command delocate-wheel --require-archs arm64 -w /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-run-8booo2i3/cp310-macosx_arm64/repaired_wheel -v /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-run-8booo2i3/cp310-macosx_arm64/built_wheel/tiledbsoma-1.1.1.post12.dev223375155-cp310-cp310-macosx_11_0_arm64.whl failed with code 1. None
```

This would seem to indicate that `tiledbsoma/libtiledbsoma.dylib` may not actually be compiled for the `arm64` arch.  See https://pypi.org/project/delocate/#checking-required-architectures for general info.

**Comment by atolopko-czi** [comment]
After setting `CMAKE_OSX_ARCHITECTURES: arm64`, it appears `tiledbsoma.dylib` is being complied to the correct arch, but is then attempting to link in an x86 `tiledb.dylib`, which fails:

```
2023-03-21T15:16:55.0553470Z   [ 33%] Linking CXX shared library libtiledbsoma.dylib
2023-03-21T15:16:55.1095090Z   ld: warning: ignoring file /private/var/folders/24/8k48jl6d249_n_qfxwsl6xvm0000gn/T/cibw-sdist-pwice1ar/tiledbsoma-1.1.1.post15.dev181249274/dist_links/build/externals/install/lib/libtiledb.dylib, building for macOS-arm64 but attempting to link with file built for macOS-x86_64
```

**Comment by bkmartinjr** [comment]
> Can you explain this issue's dependency on #494?

I believe @johnkerl added that comment - he should be able to provide info.

**Comment by johnkerl** [comment]
@atolopko-czi @bkmartinjr if the comment in question is

> N.B. Needs to be sequenced after https://github.com/single-cell-data/TileDB-SOMA/issues/494 lands

then IIRC I had in mind from someone/somewhen that we needed Conda things merged into main first. But they're already there so it's a moot point.

I have no deep knowledge of Conda; all I could have possible been doing at the time I wrote that was to give one expert a heads-up that (as far as I knew/understood) their work depended on another's work.

I have no more insight to offer; my apologies.

**Comment by atolopko-czi** [comment]
I pushed this forward, but did not solve it.  

Fixed the error:
```cibuildwheel: No build identifiers selected: BuildSelector(build_config='cp3*-macosx_arm64', skip_config='', requires_python=<SpecifierSet('>=3.7')>, prerelease_pythons=False)
```

But next error to troubleshoot is:
```
 ld: symbol(s) not found for architecture arm64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by johnkerl** [comment]
@atolopko-czi what was the command that led to this error?

```
 ld: symbol(s) not found for architecture arm64
    clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Comment by atolopko-czi** [comment]
> @atolopko-czi what was the command that led to this error?
> 
> ```
>  ld: symbol(s) not found for architecture arm64
>     clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```

The logs that led up the error: https://github.com/single-cell-data/TileDB-SOMA/actions/runs/4480707236/jobs/7876392038?pr=810#step:4:602

Also, when looking back, I realized the `kerl/m1-wheels` branch is no longer around, so I re-pushed it.

**Comment by johnkerl** [comment]
So at first blush it looks like we just haven't build `libtiledbsoma`, e.g. with `scripts/bld --prefix=/usr/local`

For R, at present this is explicit: https://github.com/single-cell-data/TileDB-SOMA/blob/main/.github/workflows/r-ci.yml#L40-L41

As of WIP #1361, we'll pull from source to build `libtiledbsoma`

This being Python -- `setup.py` already invokes `bld` so I'm not sure what was missed

As a next step (acknowledging this isn't super-critical for the current sprint) I think we should rebase onto tip of `main`. (I can do that.)

Also I have tiledbsoma working fine on my M1 laptop so I don't believe there are processor-specific issues -- just a matter of patching it through `setup.py`.

**Comment by eddelbuettel** [comment]
> we'll pull from source 

If and only if no system is found. Also PR (with approval count at one), not WIP.

**Comment by johnkerl** [comment]
#1385 is merged
